---
title: "Misc"
output:
  md_document:
    variant: markdown_github
    toc: true
  html_document:
    toc: true
    toc_float: true
---
```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt <- date()

# Packages
library(tidyverse)
library(Seurat)


# Paths
wd <- "/hpc/group/chsi-mic-2022"
intermeddir <- file.path(wd, "intermed")
```

## Load lis with Seurat objects
```{r loaddata}
infile <- file.path(intermeddir, "seulist-drc.rds")
tools::md5sum(infile)
seulist.drc <- readRDS(infile)
```
## basename and gsub

```{r}
filename <- "/hpc/group/chsi-mic-2022/data/Christian2021CellReports//disTrm_10x.h5"
tools::md5sum(filename)
bname <-basename(filename)
bname
gsub(".h5", "", bname)
```
## Extract Tem Seurat object
```{r tem}
tem <- seulist.drc[["Tem"]]
tem
```
## Extract gene names and cell ids
```{r genenames}
genenames <- rownames(tem)
genenames[1:10]
```
```{r cellids}
cellids <- colnames(tem)
cellids[1:10]
```
```{r confirm}
# Confirm that the number of genes and cells 
tem
length(cellids)
length(genenames)
```

## SCTransform
```{r }
# before
tem
```
```{r }
# Apply vst tramsformation
tem <- SCTransform(tem)
```
```{r after}
# Object after transformation
tem
```
```{r assays}
# Listing of assays
tem@assays
```

```{r activeassay}
tem@active.assay
```
```{r sctslots}
slotNames(tem@assays$SCT)
```
```{r sctdata}
dim(tem@assays$SCT@data)
```

## Extract annotation data from a GTF

```{r gtffile}
gtffile <- '/hpc/group/chsi-mic-2022/annotation/refdata-gex-GRCh38-2020-A/genes/genes.gtf'
tools::md5sum(gtffile)
```
```{r pkg}
library(rtracklayer)
library(plyranges)
```
```{r importgtf}

gtf <- import(gtffile)
```

```{r inspect}
gtf
```
```{r genes}
gtf %>% filter(type == "gene" & gene_type == "protein_coding") %>% as_tibble()
```
```{r dupgenes}

gtf %>% plyranges::filter(type == "gene") %>% 
  as_tibble() %>% 
  select(seqnames, start, end, gene_id, gene_name) %>% 
  dplyr::group_by(gene_name) %>% 
  dplyr::filter(dplyr::n()>1)
```






```{r sessioninfo}
sessionInfo()
```
