---
title: "QC scRNAseq"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---
```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt<-date()

# Libraries
library(tidyverse)
library(Seurat)

# Input paths
cr.outs <- "/hpc/group/chsi-mic-2022/data/pbmc8k/pbmc8k/outs"
intermeddir <- "/hpc/group/chsi-mic-2022/intermed"

# Output paths
scratchdir <- "/work"
username <- Sys.info()[["user"]]
outdir <- file.path(scratchdir,username,"output")
#dir.create(outdir) # uncomment if not already there
```

**Goal of this workshop:** Learn how to QC scRNAseq data

**What's covered in this workshop:**
- Convert cellranger output to a Seurat object
- Calculate mitrochondiral gene percentage
- Learn data quality attributes
- Exclude data from low-quality cells

**Data source:** 
scRNA-seq dataset of 8K human peripheral blood mononuclear cells (PBMCs) freely available from 10X Genomics (https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k)

## Convert cellranger output to a Seurat object

We can use code like this to load the cell ranger output and store it in a Seurat object.

```{r}
Seurat::Read10X_h5(file.path(cr.outs, "filtered_feature_bc_matrix.h5")) %>% 
  Seurat::CreateSeuratObject() -> pbmc8k
pbmc8k
```


## Calculate mitrochondiral gene percentage

First, take look at the existing cell-level metadata in one of the seurat objects
```{r}
head(pbmc8k@meta.data)
```

orig.ident = sample name
nCount_RNA = number of reads per cell
nFeature_RNA = number of gene features per cell

The rownames are gene symbols. Investigating the names a bit more, the mitochondrial genes can be identified by the prefix "mt-"
```{r}
curr.count <- pbmc8k@assays$RNA@counts
rownames(curr.count)[1:10]
rownames(curr.count)[grepl("^MT-", rownames(curr.count))] # these look like mitochondrial genes!
```

Add the mitochondrial percentage for 1 seurat object
```{r}
head(pbmc8k@meta.data)
pbmc8k[["percent.mt"]] <- PercentageFeatureSet(pbmc8k, pattern = "^MT-")
head(pbmc8k@meta.data)
```


## Learn data quality attributes, Initial QC

Number of molecules aka reads detected per cell (nCount)

- A very low number of reads per cell could indicate a sequencing failure.
- A very high number of reads per cell could indicate more than one cell was actually sequenced.

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = orig.ident, y = nCount_RNA))+
	geom_jitter(shape = 16, position = position_jitter(0.2))+
	geom_violin(trim = F, alpha = 0.7) +
  labs(x = "Group", y = "nCount") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, hjust=1),
        strip.text = element_text(size = 6))
```

Number of gene features detected per cell (nFeature)

- A very low number of gene features per cell could indicate a library prep or sequencing failure.

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = orig.ident, y = nFeature_RNA))+
	geom_jitter(shape = 16, position = position_jitter(0.2))+
	geom_violin(trim = F, alpha = 0.7)+
  labs(x = "Group", y = "nFeature") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, hjust=1),
        strip.text = element_text(size = 6))
```

Percentage of mitochondrial genes per cell (MT%)

- A high percentage of mitochondrial genes (MT%) indicates a cell may be dead or dying based on the expectation that, if a cell is ruptured, non-MT genes will leak out first and increase the relative abundance of MT genes.

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = orig.ident, y = percent.mt)) +
  geom_jitter(shape=16, position=position_jitter(0.2))+
	geom_violin(trim=FALSE, alpha=0.7)+
  scale_y_continuous(limits = c(0, 10)) +
  labs(x = "Group", y = "MT%") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, hjust=1),
        strip.text = element_text(size = 6)) +
  geom_hline(yintercept = 5, linetype = 2)
```

Correlation between nCounts and nFeatures, colored by MT%

- The number of gene features detected in a cell (nFeatures) tends to increase with library size (nCounts). Divergence from this correlation could indicate low-quality data, e.g. often observed in high MT% cells.

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = nCount_RNA, y = nFeature_RNA, 
                color = percent.mt))+
	geom_point(size = 1)+
	labs(x = "nCount", y = "nFeature", color = "MT%") +
  theme_classic() +
  geom_hline(yintercept = 200, linetype = 2)
```


## Exclude data from low-quality cells

"Low-quality cells that had less than 200 expressed genes and more than 5% mitochondrial genes were filtered out."


### nCount

```{r}
cl <- 6 # min number of total reads per cell (nCount is analogous to library size or total number of reads)
ch <- 11 # max number of total reads per cell
```

Filter out cells with log(nCount) $>$ `ch` \& $<$ `cl`

- Note: Different y-axis scales

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = log(nCount_RNA))) +
  geom_histogram(bins = 50) +
  geom_vline(xintercept = c(cl, ch), 
             color = "blue", linetype = "dashed") +
  labs(x = "log(nCount)", y = "Frequency") +
  scale_x_continuous(breaks = seq(0, 10, 1)) +
  theme_classic()
```


### nFeature

```{r}
fl <- 5.5 # min number of gene features per cell
```

Filtering out cells with log(nFeature) $<$ `fl`

- Note: Different y-axis scales

```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = log(nFeature_RNA))) +
  geom_histogram(bins = 100) +
  geom_vline(xintercept = fl, 
             color = "blue", linetype = "dashed") +
  scale_x_continuous(breaks = c(seq(0, 20, 1), fl)) +
  labs(x = "log(nCount)", y = "Frequency") +
  theme_classic()
```


### MT%

```{r}
tmp <- lapply(seq(1, 40, 1), function(v){
  
  pbmc8k@meta.data %>%
    group_by(orig.ident) %>%
    summarize(threshold = v, 
              ncells_kept = sum(percent.mt < v),
              ncells_filtered = sum(percent.mt >= v))
}) %>% 
  bind_rows() %>%
  pivot_longer(cols = c("ncells_filtered", "ncells_kept"),
               names_to = "type",
               values_to = "ncells") 

ml <- 5
ggplot(tmp, aes(x=threshold, y=ncells, colour=type))+
	geom_line() + 
  geom_vline(xintercept = ml, color = "blue", linetype = "dashed") + 
  scale_x_continuous(breaks = seq(0, 40, by = 5)) +
	labs(x="Thresholds for MT%", y="Number of Cells")+
  facet_wrap(~orig.ident, nrow = 3, 
             scales = "free_y")+
  theme_classic()
```

Filtering out cells with MT% $>$ `ml`
There aren't any of these cells because they have already been filtered.
```{r}
pbmc8k@meta.data %>%
  ggplot(aes(x = percent.mt)) +
  geom_histogram(bins = 100) +
  geom_vline(xintercept = ml,
             color = "blue", linetype = "dashed") +
  # scale_x_continuous(breaks = c(seq(2, 14, 2), fl)) +
  labs(x = "MT%", y = "Frequency") +
  theme_classic()
```

### Do the filtering

Pre

```{r}
nrow(pbmc8k@meta.data)
```

Filter 

```{r, echo = T}
seu_filt <- function(seuobj){
  seuobj_filt <- subset(seuobj, 
         nCount_RNA > exp(cl) & 
         nCount_RNA < exp(ch) &
         nFeature_RNA > exp(fl) &
         percent.mt < ml)
  return(seuobj_filt)
}

pbmc8k.filt <- seu_filt(seuobj = pbmc8k)
```

Post

```{r}
nrow(pbmc8k.filt@meta.data)
```

Number of cells removed by filtering
```{r}
nrow(pbmc8k@meta.data) - nrow(pbmc8k.filt@meta.data)
```




## Final QC

nCount

```{r}
pbmc8k.filt@meta.data %>%
  ggplot(aes(x = orig.ident, y = nCount_RNA))+
	geom_jitter(shape = 16, position = position_jitter(0.2))+
	geom_violin(trim = F, alpha = 0.7) +
  labs(x = "Group") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, hjust=1))
```

nFeature

```{r}
pbmc8k.filt@meta.data %>%
  ggplot(aes(x = orig.ident, y = nFeature_RNA)) +
	geom_jitter(shape = 16, position = position_jitter(0.2))+
	geom_violin(trim = F, alpha = 0.7)+
  labs(x = "Group") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 30, hjust=1))
```

MT%

```{r}
pbmc8k.filt@meta.data %>%
  ggplot(aes(x = orig.ident, y = percent.mt)) +
  geom_jitter(shape=16, position=position_jitter(0.2))+
	geom_violin(trim=FALSE, alpha=0.7)+
  labs(x = "Group", y = "MT%") +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 30, hjust=1))
```

Correlation between nCounts and nFeatures, colored by MT%

```{r}
pbmc8k.filt@meta.data %>%
  ggplot(aes(x = nCount_RNA, y = nFeature_RNA, 
                color = percent.mt))+
	geom_point(size = 1)+
	labs(x = "nCount", y = "nFeature", color = "MT%") +
  theme_classic()
```


## Session Info
```{r echo=T}
sessionInfo()
print(paste("Start Time:  ",stdt))
print(paste("End Time:  ",date()))
```

