---
title: "Load data into R and transform"
output:
  md_document:
    variant: markdown_github
    toc: true
  html_document:
    toc: true
    toc_float: true
---
```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt<-date()

# Libraries
library(tidyverse)
library(Seurat)
library(cowplot)

# Paths
wd <- "/hpc/group/chsi-mic-2022"
datadirs <- list.files(path = file.path(wd, "data","Christian2021CellReports"), pattern = "_10x", full.names = T)
datadirs <- datadirs[!grepl(".h5", datadirs)]
intermeddir <- file.path(wd, "intermed")
outdir <- file.path("/work",Sys.info()[["user"]],"output")
outdir
```

**Goal of this workshop:** Learn how to load cellranger output into R and prepare QC'd data for downstream analyses

**What's covered in this workshop:**
- Load data into R as a Seurat object
- Verify that the data have been QC'd
- Transform gene counts

**Data source:** 
Christian et al. 2021 Cell Reports (https://doi.org/10.1016/j.celrep.2021.109118)

Note that the Seurat - Guided Clustering Tutorial (https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) includes a lot of overlapping information.

**METHOD DETAILS, Single-cell RNA-seq analysis, page e4**

- Pre-processing: "Raw short reads were demultiplexed, filtering and mapped to mouse genome GRCm38/mm10 using cellranger v2.02."

- QC: "The gene count matrices from cellranger were subjected to quality control, pre-processing and clustering using the R Seurat 2.3.4 package (Butler et al., 2018). Low-quality cells that had less than 200 expressed genes and more than 5% mitochondrial genes were filtered out."

- Analysis: "Gene counts were scaled to total gene expression and percentage of mitochondrial genes with a scaling factor of 10,000, and then log-transformed. The high dimensional data for each sample were reduced by PCA and t-Distributed Stochastics Neighbor Embedding (tSNE). We used the FindCluster function to group clusters in each sample with a resolution of 0.6. Differential expressed genes (DEGs) were identified using the Wilcoxon rank-sum test."

## 1. Load data into R as a Seurat object

We can use code like this to load the cell ranger output and store it in a Seurat object.

```{r}
readin_10x_to_SO <- function(sample, path, min.cells = 0, min.genes = 0){
  Seurat::Read10X(data.dir = path) %>%
    Seurat::CreateSeuratObject(project = sample, 
                               min.cells=min.cells, min.genes=min.genes) -> seu.obj
  return(seu.obj)
}
```

Load the four 10x datasets that correspond to T cells sampled in and outside of the tumor. See Fig 2A for tSNE plots of each sample type.

- Tem = effector memory in tumor
- Tcm = central memory in tumor
- Trm = resident memory in tumor
- disTrm = resident memory T-cells outside of the tumor 

```{r}
sampleNames <- gsub("_10x","", basename(datadirs))
sampleNames

seulist <- list()
for(i in 1:length(sampleNames)){
  seulist[[i]] <- readin_10x_to_SO(sample = sampleNames[i], path = datadirs[i])
}
names(seulist) <- sampleNames
seulist
```

## 2. Verify data have been QC'd

*METHOD DETAILS, Single-cell RNA-seq analysis, page e4*: "Low-quality cells that had less than 200 expressed genes and more than 5% mitochondrial genes were filtered out."

The rownames are gene symbols. Investigating the names a bit more, the mitochondrial genes can be identified by the prefix "mt-"
```{r}
tem <- seulist$Tem
curr.count <- tem@assays$RNA@counts
rownames(curr.count)[1:10]
rownames(curr.count)[grepl("^mt-", rownames(curr.count))] # these look like mitochondrial genes!
```

Add the mitochondrial percentage for all seurat objects in our list
```{r}
add_mt <- function(so){
  so[["percent.mt"]] <- PercentageFeatureSet(so, pattern = "^mt-")
  return(so)
}

seulist %>%
  map(~add_mt(.x)) -> seulist.mt

seulist.mt %>%
  map(~head(.x@meta.data))

seulist <- seulist.mt
```

Examine the distribution of nFeature and nCount for each dataset
```{r}
seulist %>%
  map_dfr(~.x@meta.data, .id = "seuobj") %>%
  ggplot(aes(x = nCount_RNA, y = nFeature_RNA, 
                color = percent.mt))+
	geom_point(size = 1)+
	facet_wrap(~ seuobj, nrow = 2)+
	labs(x = "nCount", y = "nFeature", color = "MT%") +
  theme_classic() +
  geom_hline(yintercept = 200, linetype = 2)
```
Are there any cells with less than 200 expressed genes?
```{r}
seulist %>%
  map_dfr(~.x@meta.data, .id = "seuobj") %>%
  ggplot(aes(x = nFeature_RNA))+
	geom_histogram() +
	facet_wrap(~ seuobj, nrow = 2)+
  theme_classic() +
  geom_vline(xintercept = 200, linetype = 2)
```
Summarize as a table
```{r}
seulist %>%
  map_dfr(~.x@meta.data, .id = "seuobj") %>%
  group_by(seuobj) %>%
  summarize(min = min(nFeature_RNA),
            mean = mean(nFeature_RNA),
            max = max(nFeature_RNA))
```

Are there any cells with more than 5% mitochondrial genes?
```{r}
seulist %>%
  map_dfr(~.x@meta.data, .id = "seuobj") %>%
  ggplot(aes(x = percent.mt))+
	geom_histogram() +
	facet_wrap(~ seuobj, nrow = 2)+
  theme_classic() +
  geom_vline(xintercept = 5, linetype = 2)
```
Summarize as a table
```{r}
seulist %>%
  map_dfr(~.x@meta.data, .id = "seuobj") %>%
  group_by(seuobj) %>%
  summarize(min = min(percent.mt),
            mean = mean(percent.mt),
            max = max(percent.mt))
```

## 3. Transform gene counts

*METHOD DETAILS, Single-cell RNA-seq analysis, page e4*: "Gene counts were scaled to total gene expression and percentage of mitochondrial genes with a scaling factor of 10,000, and then log-transformed."

First, take a look at how data are structured inside the filtered seurat object
```{r}
tem <- seulist$Tem

Assays(tem)
tem@assays
```

There is 1 assay called "RNA". Inside this, there are 3 slots for matrices named "counts", "data", and "scale.data". Each matrix consists of rows = features and columns = cells.
```{r}
curr.counts <- tem@assays$RNA@counts
dim(curr.counts) # 12571 features as rows and 1457 cells as columns
colnames(curr.counts)[1:10]
rownames(curr.counts)[1:10]

curr.data <- tem@assays$RNA@data
dim(curr.data) # 12571 features as rows and 1457 cells as columns
colnames(curr.data)[1:10]
rownames(curr.data)[1:10]

curr.scale.data <- tem@assays$RNA@scale.data
dim(curr.scale.data) # empty
```

Compare how the "counts" and "data" values differ. Pull out values for a single cell and plot.
```{r}
i <- 1
colnames(curr.counts)[i] # cell identifier
df <- data.frame(count = curr.counts[,i], data = curr.data[,i], feature = row.names(curr.counts), row.names = NULL)
head(df)

ggplot(df, aes(x = count, y = data)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey")
```

Normalize the counts and re-examine the data slots. Note that only the values in the "data" slot have changed.
```{r}
tem.s <- NormalizeData(object = tem, normalization.method = "LogNormalize", scale.factor = 10000)

s.counts <- tem.s@assays$RNA@counts
dim(s.counts) # 12571 features as rows and 1457 cells as columns
s.counts[1:10,1:10]
#colnames(s.counts)[1:10]
#rownames(s.counts)[1:10]

s.data <- tem.s@assays$RNA@data
dim(s.data) # 12571 features as rows and 1457 cells as columns
s.data[1:10,1:10]
#colnames(s.data)[1:10]
#rownames(s.data)[1:10]

s.scale.data <- tem.s@assays$RNA@scale.data
dim(s.scale.data) # empty

## plot
i <- 1
colnames(s.counts)[i] # cell identifier
df <- data.frame(count = s.counts[,i], data = s.data[,i], feature = row.names(s.counts), row.names = NULL)
ggplot(df, aes(x = count, y = data)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey")
```

Normalize each of the seurat objects in the list.

Here is an example of how to do this with a for-loop
```{r}
sampleNames <- names(seulist)
seulist.n <- list()
for(i in 1:length(sampleNames)){
  seulist.n[[i]] <- NormalizeData(object = seulist[[i]], normalization.method = "LogNormalize", 
                                  scale.factor = 10000)
}
names(seulist.n) <- sampleNames
```

Here is an example of how to this using purrr::map and pipes
```{r}
# seulist %>%
#   map(~NormalizeData(object = .x, normalization.method = "LogNormalize", scale.factor = 10000)) -> seulist.n
```




