---
title: "MIC-2022 Workshop: single-cell clustering and annotation"
output: html_notebook
---

Before we start, we set up the R environment.

```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt <- date()

# Libraries
library(tidyverse)
library(Seurat)
library(cowplot)

# Input paths
#wd <- "~/Documents/Teaching/chsi-mic-2022"
wd <- "/hpc/group/chsi-mic-2022"
intermeddir <- file.path(wd, "intermed")

# Output paths
scratchdir <- "~/Documents/Temporary/scratch"
# scratchdir <- "/work"
# username <- Sys.info()[["user"]]
# outdir <- file.path(scratchdir,username,"output")
# dir.create(outdir) # uncomment if not already there
```

## Overview

**Data source:** Christian et al. 2021 Cell Reports (https://doi.org/10.1016/j.celrep.2021.109118)

**Goal of this workshop:** Illustrate the clustering and cell-type annotation procedures. 

Note that the Seurat - [Guided Clustering Tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) includes a lot of overlapping information.

**METHOD DETAILS, Single-cell RNA-seq analysis, page e4**
- Pre-processing: "Raw short reads were demultiplexed, filtering and mapped to mouse genome GRCm38/mm10 using cellranger v2.02."
- QC: "The gene count matrices from cellranger were subjected to quality control, pre-processing and clustering using the R Seurat 2.3.4 package (Butler et al., 2018). Low-quality cells that had less than 200 expressed genes and more than 5% mitochondrial genes were filtered out."
- Analysis: "Gene counts were scaled to total gene expression and percentage of mitochondrial genes with a scaling factor of 10,000, and then log-transformed. The high dimensional data for each sample were reduced by PCA and t-Distributed Stochastics Neighbor Embedding (tSNE). We used the FindCluster function to group clusters in each sample with a resolution of 0.6. Differential expressed genes (DEGs) were identified using the Wilcoxon rank-sum test."

**What's covered in this workshop:**
- Highly variable gene identification
- Dimension reduction
- Clustering and visualization
- Cell-type specific marker gene detection
- How to generate the figures for the paper?

## Load data

First, we load the normalized UMI count single-cell data.

```{r load data}
seulist <- readRDS(file.path(intermeddir, "seulist_drc.rds"))
ls()
seulist
tem <- seulist[["Tem"]]
```
## Highly variable gene (HVG) identification

**First, find the top 2000 most variable gene features.** This function finds the features, stashes them inside the seurat object, then returns the seurat object.

```{r}
tem.vf <- FindVariableFeatures(tem, selection.method = "vst", nfeatures = 2000)
```

You can extract the variable features from the seurat object using these methods.
```{r}
VariableFeatures(tem.vf)[1:10] # option 1
tem.vf@assays$RNA@var.features[1:10] # option 2

# note that this slot is empty in the seurat object prior to the FindVariableFeatures() command
tem@assays$RNA@var.features
```

In the `FindVariableFeatures()` function, we used `vst` as the selection method. Intermediate data associated with that selection method is stored inside the seurat object here:
```{r}
head(tem.vf@assays$RNA@meta.features)
```

Seurat has a function to generate a plot to show how variable these "variable features" are.
```{r}
top10 <- head(VariableFeatures(tem.vf), 10)
plot1 <- VariableFeaturePlot(tem.vf)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 
plot2
```

## Dimension reduction

*METHOD DETAILS, Single-cell RNA-seq analysis, page e4*: "The high dimensional data for each sample were reduced by PCA..."

Specifically, shift the expression of each gene, so that the mean expression across cells is 0 and the variance across cells is 1. If we didn't do this, highly-expressed and highly-variable genes would dominate the outcome of down-stream analyses.

```{r}
all.genes <- rownames(tem.vf)
tem.sAll <- ScaleData(tem.vf, features = all.genes) 
tem.sVF <- ScaleData(tem.vf) 

# note that if you do not specify features as all.genes, the default behavior is to only use the 2000 variable features downstream
```

Examine the scaled data.
```{r}
# prior to scaling
dim(tem.vf@assays$RNA@scale.data) # empty

# after scaling with all genes
dim(tem.sAll@assays$RNA@scale.data) # 12571 features as rows and 1457 cells as columns

# after scaling with the 2000 variable features
dim(tem.sVF@assays$RNA@scale.data) # 2000 features as rows and 1457 cells as columns
```

We perform PCA on the scaled data. 

```{r}
tem.pAll <- RunPCA(tem.sAll, features = all.genes) # this takes a minute or so...
tem.pVF <- RunPCA(tem.sVF)
# again, note that the default behavior is to use only the variables to compute the PCA
```


The `RunPCA()` function automatically spits out information about how features loaded into PC space. But like most Seurat functions, you will need to look inside the object to get details.

Here is where you will find detailed info about the reduction you performed
```{r}
tem.pVF@reductions # the seurat object keeps track of reductions you have performed here
names(tem.pVF@commands) # more broadly, there is a log of all commands stored here
tem.pVF@commands$RunPCA.RNA # with even more details
```

Next, we can use the visualization tools to the choose the proper number of PCs.

```{r}
VizDimLoadings(tem.pVF, dims = 1:2, reduction = "pca")
DimPlot(tem.pVF, reduction = "pca")
DimHeatmap(tem.pVF, dims = 1, cells = 500, balanced = TRUE)
```

Check the first 15 PCs
```{r, fig.width=20, fig.height=30}
DimHeatmap(tem.pVF, dims = 1:15, cells = 500, balanced = TRUE)
```

To refine the number of dimensions, we can use...

**Option 1:**
A resampling test inspired by the JackStraw procedure developed by the Seurat authors. The idea is.."to permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value features....The `JackStrawPlot()` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line)."(https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).

```{r}
#tem.js <- JackStraw(tem.pVF, num.replicate = 100) # this takes a minute or two, so we can just load a version that we previously created, cooking-show style.

# Students will not be able to run the saveRDS command because 
# everything the data/ and intermed/ folders will need to be write-protected
#saveRDS(tem.js, file = file.path(intermeddir, "tem_js.rds"))
tem.js <- readRDS(file.path(intermeddir, "tem_js.rds"))

tem.js <- ScoreJackStraw(tem.js, dims = 1:20)
JackStrawPlot(tem.js, dims = 1:20)
```

There seems to be a drop-off in significance somewhere between PC5 and PC10 but it is difficult to tell with the colors. This suggests we should use ~10 dimensions.

**Option 2:**
We can also just take a look at the percentage of variance explained by each additional PC. We don't want to bother including PCs that explain very little additional variance so we can look for the "elbow" where the variance explained (or inversely, the standard deviation) levels-off. Seurat's tool, `ElbowPlot()` displays standard deviation on the y-axis.
```{r}
ElbowPlot(tem.pVF)
```
Here we can see an "elbow" that forms around PC10 to PC15, suggesting that we should to use at least that many dimensions and, beyond that, we start seeing diminishing returns.

## Clustering and visualization

*METHOD DETAILS, Single-cell RNA-seq analysis, page 4e*: " ..and t-Distributed Stochastics Neighbor Embedding (tSNE). We used the `FindCluster()` function to group clusters in each sample with a resolution of 0.6."

```{r}
tem.pVF <- FindNeighbors(tem.pVF, dims = 1:10)
tem.pVF <- FindClusters(tem.pVF, resolution = 0.6)
```

The cluster IDs are stored here.
```{r}
Idents(tem.pVF)[1:10]
```

After finding clusters, project into 2D for visualization using tSNE.
```{r}
tem.pVf.tsne <- RunTSNE(tem.pVF, dims = 1:10) 
# note that the seurat authors suggest using the same PCs as input to the clustering analysis
DimPlot(tem.pVf.tsne, reduction = "tsne") + coord_equal()
```


Another option is to do this with UMAP instead of tSNE. They should look pretty similar.
```{r}
tem.pVf.umap <- RunUMAP(tem.pVF, dims = 1:10)
DimPlot(tem.pVf.umap, reduction = "umap") + coord_equal()
```

### Apply to the list of seurat objects

Make a function that contains that takes a seurat object as input, does all the operations we want, then returned the updated seurat object.
```{r}
dimRed_and_cluster <- function(seuobj){
  
  #seuobj <- FindVariableFeatures(seuobj, selection.method = "vst", nfeatures = 2000)
  all.genes <- rownames(seuobj)
  seuobj <- ScaleData(seuobj, features = all.genes) 
  seuobj <- RunPCA(seuobj, features = all.genes)
  seuobj <- FindNeighbors(seuobj, dims = 1:10)
  seuobj <- FindClusters(seuobj, resolution = 0.6)
  seuobj <- RunTSNE(seuobj, dims = 1:10)
  
  return(seuobj)

}
```

Apply that function to each seurat object in the list of objects that we normalized
```{r}
#seulist.n %>%
#  map(~dimRed_and_cluster(.x)) -> seulist.drc
# this takes a minute or two, so we can just load a version that we previously created, cooking-show style.

# Students will not be able to run the saveRDS command because 
# everything the data/ and intermed/ folders will need to be write-protected
#saveRDS(seulist.drc, file = file.path(intermeddir, "seulist_drc.rds"))
seulist.drc <- readRDS(file.path(intermeddir, "seulist_drc.rds"))
```

### Examine the tSNE plots (Fig2A). 

These should resemble plots in Fig2, page 5.

**Fig 2A**. They are similar but have different orientations and numbers of clusters.
```{r}
DimPlot(seulist.drc[["Tem"]], reduction = "tsne") + ggtitle("Tem") + coord_equal()
# Clusters 0, 1, 3 correspond to p1?
# Cluster 3 corresponds to p2?
# Cluster 4 corresponds to p3?
# Cluster 5 corresponds to p4?

DimPlot(seulist.drc[["Tcm"]], reduction = "tsne") + ggtitle("Tcm") + coord_equal()

DimPlot(seulist.drc[["Trm"]], reduction = "tsne") + ggtitle("Trm") + coord_equal()

DimPlot(seulist.drc[["disTrm"]], reduction = "tsne") + ggtitle("disTrm") + coord_equal()
```

Overlay the expression of 1 gene on top of the Tem cells in tSNE space as shown in **Fig 2B**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Ifng")) + coord_equal()
#FeaturePlot(seulist.drc[["Tem"]], features = c("GzmB"))
#The following requested variables were not found: GzmB????
FeaturePlot(seulist.drc[["Tem"]], features = c("Klrc1")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Nkg7")) + coord_equal()
```

Overlay ... as shown in **Fig 2C**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Runx3")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Id2")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Id3")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Bcl2")) + coord_equal()
```

Overlay ... as shown in **Fig 2D**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Ccnb2")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Cdk1")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Mki67")) + coord_equal()
```


## Cell-type specific marker gene detection

*METHOD DETAILS, Single-cell RNA-seq analysis, page 4e*: "Differential expressed genes (DEGs) were identified using the Wilcoxon rank-sum test."

#### Identify Tem cluster biomarkers (Fig2B-D)

```{r}
tem <- seulist.drc[["Tem"]]
tem@meta.data %>%
  pull(seurat_clusters) %>% unique()
```

For the purpose of trying to reproduce the manuscript results, we will assume that the authors use defaults. Here are a few default arguments to be aware of:

**min.pct = 0.1** : Do not test genes that make up fewer than 0.1 fraction (10% ?) of the total reads in either of the populations tested. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1 -- **NOTE: Marissa tried to re-write this so that it is clearer, please check**

**max.cells.per.ident = Inf** : "This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top" (https://satijalab.org/seurat/articles/pbmc3k_tutorial.html)

**logfc.threshold = 0.25** "Limit testing to genes which show, on average, at least X-fold difference (log-scale) between the two groups of cells. Default is 0.25 Increasing logfc.threshold speeds up the function, but can miss weaker signals." (Seurat help)

**test.use = "wilcox** Denotes which test to use; see Seurat help for alternatives.

**min.cells.group = 3** "Minimum number of cells in one of the groups" (Seurat help)

**ident.2** "A second identity class for comparison; if NULL, use all other cells for comparison" (Seurat help)

Cluster 0
```{r}
cluster0.markers <- FindMarkers(tem, ident.1 = 0)
cluster0.markers %>%
  arrange(-avg_log2FC) %>%
  head(n = 5)
```
Check the expression of the feature genes.
```{r}
FeaturePlot(tem, features = c("Bcl2")) + coord_equal()
```


