---
title: "Trajectory analysis with monocle"
output: html_notebook
---
```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt <- date()

# Libraries
library(tidyverse)
library(Seurat)
library(monocle)

# Paths
wd <- "/hpc/group/chsi-mic-2022"
intermeddir <- file.path(wd, "intermed")
outdir <- file.path("/work",Sys.info()[["user"]],"output")
#outdir
```

**Goal of this workshop:** Learn to analyze cell trajectories with monocle

**What's covered in this workshop:**

- Merge Seurat objects
- Create a Monocle object
- Run trajectory analysis
- Test for DE over pseudotime

Note that the Seurat - [Guided Clustering Tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) includes a lot of overlapping information.

**Data source:** Christian et al. 2021 Cell Reports (https://doi.org/10.1016/j.celrep.2021.109118)

**METHOD DETAILS, Single-cell RNA-seq analysis, page e4**
- Analysis: "...."

## Load data

Load Seurat objects into R

```{r}
seulist <- readRDS(file.path(intermeddir, "seulist_drc.rds"))

trm <- seulist[["Trm"]]
disttrm <- seulist[["disTrm"]]
tem <- seulist[["Tem"]]
```

## 1. Merge Seurat objects

Merge the seurat objects together so that there is 1 object that includes all samples.

Since we will be merging the cells, it is useful to add a prefix to the cell names that identify which sample the cell came from. This information can also be stored in the metadata.

```{r}
Cells(trm)[1:10]
trm.rn <- RenameCells(trm, add.cell.id = 'Trm')
Cells(trm.rn)[1:10]

# do this for the remaining samples that will be merged
tem.rn <- RenameCells(tem, add.cell.id='Tem')
disttrm.rn <- RenameCells(disttrm, add.cell.id='MucosalTrm')
```

Merge Trm and MucosalTrm cells
```{r}
comb <- merge(trm.rn, disttrm.rn)

comb@meta.data %>%
  group_by(orig.ident) %>%
  summarize(ncells = length(orig.ident))
```
Add the Tem cells
```{r}
comb <- merge(comb, tem.rn)

comb@meta.data %>%
  group_by(orig.ident) %>%
  summarize(ncells = length(orig.ident))
```
Check the cell names
```{r}
dim(comb@meta.data) #2779 cells in total

Cells(comb) %>%
  str_split("_") %>%
  map_dfr(~data.frame(prefix = .x[1], orig.cellname = .x[2])) %>%
  mutate(new.cellname = Cells(comb)) -> cellname.df

cellname.df %>%
  group_by(prefix) %>%
  summarize(ncells = length(orig.cellname))
```

## 2. Create a Monocle object

```{r}
pd <- new("AnnotatedDataFrame", data = comb@meta.data)
fd <- new("AnnotatedDataFrame", data = data.frame(gene_short_name=rownames(comb),row.names = rownames(comb)))
cds <- newCellDataSet(comb@assays$RNA@counts, phenoData = pd, featureData = fd)

```

## 3. Run tranjectory analysis

Prepare expression data

```{r}
# Estimate size factors and dispersions
cds <- estimateSizeFactors(cds)
cds <- estimateDispersions(cds)

# Filtering low expressing genes
cds <- detectGenes(cds, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(cds),num_cells_expressed >= 100))
```

```{r}
# Selecting genes for pseudotime
diff_test_res <- differentialGeneTest(cds[expressed_genes,],fullModelFormulaStr = "~type",cores=5)
ordering_genes <- row.names(subset(diff_test_res, qval < 0.05))
cds <- setOrderingFilter(cds, ordering_genes)

# Reduce data dimensionality
cds <- reduceDimension(cds, max_components = 2, method = 'DDRTree')

# Order cells along the trajectory
cds <- orderCells(cds)
```

```{r}
# Plot trajectory
pdf('/hpc/group/jilab/zj/R25/monocle2/trajectory.pdf')
plot_cell_trajectory(cds,color_by = "type")
dev.off()
```

## 4. Test for DE over psuedotime
```{r}
# Differential analysis for all Il genes
marker_genes <- grep('^Il',fData(cds)$gene_short_name,value = T)
diff_test_res <- differentialGeneTest(cds[marker_genes,],fullModelFormulaStr = "~sm.ns(Pseudotime)")

# Select genes that are significant at an FDR < 5%
sig_genes <- subset(diff_test_res, qval < 0.05)
```

```{r}
# Plot individual genes
pdf('/hpc/group/jilab/zj/R25/monocle2/Ilexpression.pdf')
plot_genes_in_pseudotime(cds[sig_genes$gene_short_name,],ncol=3)
dev.off()
save.image('/hpc/group/jilab/zj/R25/monocle2/image.rda')
```