---
title: "Analyses scRNAseq"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---
```{r prereqs}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt<-date()

# Libraries
library(tidyverse)
library(Seurat)

# Input paths
wd <- "/hpc/group/chsi-mic-2022"
datadirs <- list.files(path = file.path(wd, "data"), pattern = "_10x", full.names = T)
datadirs <- datadirs[!grepl(".h5", datadirs)]
intermeddir <- file.path(wd, "intermed")

# Output paths
scratchdir <- "/work"
username <- Sys.info()[["user"]]
outdir <- file.path(scratchdir,username,"output")
#dir.create(outdir) # uncomment if not already there
```

**Data source:** 
Christian et al. 2021 Cell Reports (https://doi.org/10.1016/j.celrep.2021.109118)

**Goal of this script:** Analysis

Note that the Seurat - Guided Clustering Tutorial (https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) includes a lot of overlapping information.

**Methods overview:**

- Pre-processing: "Raw short reads were demultiplexed, filtering and mapped to mouse genome GRCm38/mm10 using cellranger v2.02. 

- QC: The gene count matrices from cellranger were subjected to quality control, pre-processing and clustering using the R Seurat 2.3.4 package (Butler et al., 2018). Low-quality cells that had less than 200 expressed genes and more than 5% mitochondrial genes were filtered out. 

- Analysis: Gene counts were scaled to total gene expression and percentage of mitochondrial genes with a scaling factor of 10,000, and then log-transformed. The high dimensional data for each sample were reduced by PCA and t-Distributed Stochastics Neighbor Embedding (tSNE). We used the FindCluster function to group clusters in each sample with a resolution of 0.6. Differential expressed genes (DEGs) were identified using the Wilcoxon rank-sum test."


## Load QC'd Seurat objects

```{r}
seulist.filt <- readRDS(file.path(intermeddir, "seulist-filt.rds"))
names(seulist.filt)
```

## Transform counts

"Gene counts were scaled to total gene expression and percentage of mitochondrial genes with a scaling factor of 10,000, and then log-transformed." (Christian et al. 2021 Cell Reports)

First, take a look at how data are structured inside the filtered seurat object
```{r}
tem <- seulist.filt$Tem

Assays(tem)
tem@assays
```

There is 1 assay called "RNA". Inside this, there are 3 slots for matrices named "counts", "data", and "scale.data". Each matrix consists of rows = features and columns = cells.
```{r}
curr.counts <- tem@assays$RNA@counts
dim(curr.counts) # 12571 features as rows and 1457 cells as columns
colnames(curr.counts)[1:10]
rownames(curr.counts)[1:10]

curr.data <- tem@assays$RNA@data
dim(curr.data) # 12571 features as rows and 1457 cells as columns
colnames(curr.data)[1:10]
rownames(curr.data)[1:10]

curr.scale.data <- tem@assays$RNA@scale.data
dim(curr.scale.data) # empty
```

Compare how the "counts" and "data" values differ. Pull out values for a single cell and plot.
```{r}
i <- 1
colnames(curr.counts)[i] # cell identifier
df <- data.frame(count = curr.counts[,i], data = curr.data[,i], feature = row.names(curr.counts), row.names = NULL)
head(df)

ggplot(df, aes(x = count, y = data)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey")
```

Normalize the counts and re-examine the data slots. Note that only the values in the "data" slot have changed.
```{r}
tem.s <- NormalizeData(object = tem, normalization.method = "LogNormalize", scale.factor = 10000)

s.counts <- tem.s@assays$RNA@counts
dim(s.counts) # 12571 features as rows and 1457 cells as columns
s.counts[1:10,1:10]
#colnames(s.counts)[1:10]
#rownames(s.counts)[1:10]

s.data <- tem.s@assays$RNA@data
dim(s.data) # 12571 features as rows and 1457 cells as columns
s.data[1:10,1:10]
#colnames(s.data)[1:10]
#rownames(s.data)[1:10]

s.scale.data <- tem.s@assays$RNA@scale.data
dim(s.scale.data) # empty

## plot
i <- 1
colnames(s.counts)[i] # cell identifier
df <- data.frame(count = s.counts[,i], data = s.data[,i], feature = row.names(s.counts), row.names = NULL)
ggplot(df, aes(x = count, y = data)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey")
```

Normalize each of the seurat objects in the list.

Here is an example of how to do this with a for-loop
```{r}
sampleNames <- names(seulist.filt)
seulist.n <- list()
for(i in 1:length(sampleNames)){
  seulist.n[[i]] <- NormalizeData(object = seulist.filt[[i]], normalization.method = "LogNormalize", 
                                  scale.factor = 10000)
}
names(seulist.n) <- sampleNames
```

Here is an example of how to this using purrr::map and pipes
```{r}
# seulist.filt %>%
#   map(~NormalizeData(object = .x, normalization.method = "LogNormalize", scale.factor = 10000)) -> seulist.n
```


## Dimension reduction and clustering

As above, we will use 1 seurat object for demonstration and then operate on the list of seurat objects at the very end.
```{r}
names(seulist.n)
tem <- seulist.n[["Tem"]]
```

### Dimension reduction

"The high dimensional data for each sample were reduced by PCA..." (Christian et al. 2021 Cell Reports)

**First, find the top 2000 most variable gene features.** This function finds the features, stashes them inside the seurat object, then returns the seurat object.
```{r}
tem.vf <- FindVariableFeatures(tem, selection.method = "vst", nfeatures = 2000)
```

You can extract the variable features from the seurat object using these methods
```{r}
VariableFeatures(tem.vf)[1:10] # option 1
tem.vf@assays$RNA@var.features[1:10] # option 2

# note that this slot is empty in the seurat object prior to the FindVariableFeatures() command
tem@assays$RNA@var.features
```

In the FindVariableFeatures() function, we used "vst" as the selection method. Intermediate data associated with that selection method is stored inside the seurat object here:
```{r}
head(tem.vf@assays$RNA@meta.features)
```

Seurat has a function to generate a plot to show how variable these "variable features" are.
```{r}
top10 <- head(VariableFeatures(tem.vf), 10)
plot1 <- VariableFeaturePlot(tem.vf)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

**Second, scale the data.** Specifically, shift the expression of each gene, so that the mean expression across cells is 0 and the variance across cells is 1. If we didn't do this, highly-expressed and highly-variable genes would dominate the outcome of down-stream analyses.
```{r}
all.genes <- rownames(tem.vf)
tem.sAll <- ScaleData(tem.vf, features = all.genes) 
tem.sVF <- ScaleData(tem.vf) 

# note that if you do not specify features as all.genes, the default behavior is to only use the 2000 variable features downstream
```

Examine the scaled data.
```{r}
# prior to scaling
dim(tem.vf@assays$RNA@scale.data) # empty

# after scaling with all genes
dim(tem.sAll@assays$RNA@scale.data) # 12571 features as rows and 1457 cells as columns

# after scaling with the 2000 variable features
dim(tem.sVF@assays$RNA@scale.data) # 2000 features as rows and 1457 cells as columns
```

**Third, perform dimension reduction with PCA.** 
```{r}
tem.pAll <- RunPCA(tem.sAll, features = all.genes) # this takes a minute or so...
tem.pVF <- RunPCA(tem.sVF)
# again, note that the default behavior is to use only the variables to compute the PCA
```

The RunPCA() function automatically spits out information about how features loaded into PC space. But like most Seurat functions, you will need to look inside the object to get details.

Here is where you will find detailed info about the reduction you performed
```{r}
tem.pVF@reductions # the seurat object keeps track of reductions you have performed here
names(tem.pVF@commands) # more broadly, there is a log of all commands stored here
tem.pVF@commands$RunPCA.RNA # with even more details
```

Here is where you can find the PCA scores and loadings
```{r}
dim(tem.pVF@reductions$pca@cell.embeddings) # 1457 cells as rows, 50 dimensions as columns
dim(tem.pVF@reductions$pca@feature.loadings) # 2000 features as rows, 50 dimensions as columns
#str(tem.pVF@reductions$pca)
```

Seurat has a few visualization tools to see how your features loaded into PC space.
```{r}
VizDimLoadings(tem.pVF, dims = 1:2, reduction = "pca")
DimPlot(tem.pVF, reduction = "pca")
DimHeatmap(tem.pVF, dims = 1, cells = 500, balanced = TRUE)
```

To refine the number of dimensions, we can use...

**option1**
A resampling test inspired by the JackStraw procedure developed by the Seurat authors. The idea is.."to permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of feature scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value features....The JackStrawPlot() function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line)."(https://satijalab.org/seurat/articles/pbmc3k_tutorial.html).

```{r}
#tem.js <- JackStraw(tem.pVF, num.replicate = 100) # this takes a minute or two, so we can just load a version that we previously created, cooking-show style.

# Students will not be able to run the saveRDS command because 
# everything the data/ and intermed/ folders will need to be write-protected
#saveRDS(tem.js, file = file.path(intermeddir, "tem_js.rds"))
tem.js <- readRDS(file.path(intermeddir, "tem_js.rds"))

tem.js <- ScoreJackStraw(tem.js, dims = 1:20)
JackStrawPlot(tem.js, dims = 1:15)
```
There seems to be a drop-off in significance somewhere between PC5 and PC10 but it is difficult to tell with the colors. This suggests we should use ~10 dimensions.

**option2**
We can also just take a look at the percentage of variance explained by each additional PC. We don't want to bother including PCs that explain very little additional variance so we can look for the "elbow" where the variance explanined (or inversely, the standard deviation) levels-off. Seurat's tool, ElbowPlot() displays standard deviation on the y-axis.
```{r}
ElbowPlot(tem.pVF)
```
Here we can see an "elbow" that forms around PC10 to PC15, suggesting that we should to use at least that many dimensions and, beyond that, we start seeing diminishing returns.

### Cell clustering

" ..and t-Distributed Stochastics Neighbor Embedding (tSNE). We used the FindCluster function to group clusters in each sample with a resolution of 0.6." (Christian et al. 2021 Cell Reports)

```{r}
tem.pVF <- FindNeighbors(tem.pVF, dims = 1:10)
tem.pVF <- FindClusters(tem.pVF, resolution = 0.6)
```

The cluster IDs are stored here
```{r}
Idents(tem.pVF)[1:10]
```

After finding clusters, project into 2D for visualization using tSNE.
```{r}
tem.pVf.tsne <- RunTSNE(tem.pVF, dims = 1:10) 
# note that the seurat authors suggest using the same PCs as input to the clustering analysis
DimPlot(tem.pVf.tsne, reduction = "tsne")
```

Another option is to do this with UMAP instead of tSNE. They should look pretty similar.
```{r}
tem.pVf.umap <- RunUMAP(tem.pVF, dims = 1:10) 
DimPlot(tem.pVf.umap, reduction = "umap")
```

### Apply to the list of seurat objects

Make a function that contains that takes a seurat object as input, does all the operations we want, then returned the updated seurat object.
```{r}
dimRed_and_cluster <- function(seuobj){
  
  #seuobj <- FindVariableFeatures(seuobj, selection.method = "vst", nfeatures = 2000)
  all.genes <- rownames(seuobj)
  seuobj <- ScaleData(seuobj, features = all.genes) 
  seuobj <- RunPCA(seuobj, features = all.genes)
  seuobj <- FindNeighbors(seuobj, dims = 1:10)
  seuobj <- FindClusters(seuobj, resolution = 0.6)
  seuobj <- RunTSNE(seuobj, dims = 1:10)
  
  return(seuobj)

}
```

Apply that function to each seurat object in the list of objects that we normalized
```{r}
# seulist.n %>%
#   map(~dimRed_and_cluster(.x)) -> seulist.drc
# this takes a minute or two, so we can just load a version that we previously created, cooking-show style.

# Students will not be able to run the saveRDS command because 
# everything the data/ and intermed/ folders will need to be write-protected
#saveRDS(seulist.drc, file = file.path(intermeddir, "seulist_drc.rds"))
seulist.drc <- readRDS(file.path(intermeddir, "seulist_drc.rds"))
```

Examine the tSNE plots. These should resemble plots in Fig2.

"Figure 2. Single-cell RNA sequencing reveals intratumor TEff/EM heterogeneity (A) Independent tSNE plots of sorted by... single-cell transcriptomes obtained from 10 tumors and
matched distant mucosa. Each dot represents a cell; each color indicates a distinct CD8+ T cell cluster.
(B–D) TEff/EM CD8+ clusters are labeled p1–p4, and the expression of (B) effector molecules and activation markers, (C) transcription factors, and (D) proliferation markers in the tumor TEFF/EM clusters are shown." Christian et al. 2021 Cell Reports

**Fig 2A**. They are similar but have different orientations and numbers of clusters.
```{r}
DimPlot(seulist.drc[["Tem"]], reduction = "tsne") + ggtitle("Tem") + coord_equal()
# Clusters 0, 1, 3 correspond to p1?
# Cluster 3 corresponds to p2?
# Cluster 4 corresponds to p3?
# Cluster 5 corresponds to p4?

DimPlot(seulist.drc[["Tcm"]], reduction = "tsne") + ggtitle("Tcm") + coord_equal()

DimPlot(seulist.drc[["Trm"]], reduction = "tsne") + ggtitle("Trm") + coord_equal()

DimPlot(seulist.drc[["disTrm"]], reduction = "tsne") + ggtitle("disTrm") + coord_equal()
```

Overlay the expression of 1 gene on top of the Tem cells in tSNE space as shown in **Fig 2B**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Ifng")) + coord_equal()
#FeaturePlot(seulist.drc[["Tem"]], features = c("GzmB"))
#The following requested variables were not found: GzmB????
FeaturePlot(seulist.drc[["Tem"]], features = c("Klrc1")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Nkg7")) + coord_equal()
```

Overlay ... as shown in **Fig 2C**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Runx3")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Id2")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Id3")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Bcl2")) + coord_equal()
```

Overlay ... as shown in **Fig 2D**.
```{r}
FeaturePlot(seulist.drc[["Tem"]], features = c("Ccnb2")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Cdk1")) + coord_equal()
FeaturePlot(seulist.drc[["Tem"]], features = c("Mki67")) + coord_equal()
```


## Differential expression

"Differential expressed genes (DEGs) were identified using the Wilcoxon rank-sum test." (Christian et al. 2021 Cell Reports)

"Figure 3. Two distinct subsets that resemble either TEMs or TCMs comprise tumor and distant mucosa TRMs...
(B) Heatmap of top 100 DEGs between the tumor TEff/EM and tumor TCMs shows the global gene expression pattern between these circulating memory T cells and active and quiescent TRMs....(E) A global enrichment of ribosome genes associates with tumor TCM and quiescent TRM subsets." (Christian et al. 2021 Cell Reports) **Note** In order to make Fig3, need to be able to distinguish between active and quiesent Trm and disTrm cells. This depends on Monocle2 analysis.

"Figure 5. CXCR6 defines a unique subpopulation of TEff/EMs ...
(B) Elevated levels of Cxcr6 in TEff/EM p4 are associated with enhanced Pdcd1 expression, in opposition to Il7r and Il18r1 expression, which are expressed in TEff/EM p2. Classical exhaustion markers Nr4a1, Lag3, and Havcr2 were upregulated in the tumor TEff/EM p4 population.
(C) Volcano plot of gene expression between tumor TEff/EM p2 (left) and TEff/EM p4 (right) shows that T cell activation markers Gzmb, Klrc1/2, and Klrd1 are upregulated in TEFF/TEM p4, indicating it is effector like.
(E) PCA analysis confirms transcriptional separation of these 2 populations. **What are the points???**
(F) Gene Ontology analysis shows that the CXCR6+ population enriches many pathways involving the cell cycle, indicating that this pathway is phenotypically exhausted but highly active." (Christian et al. 2021 Cell Reports)



