---
title: "Cellranger Demo"
output: html_notebook
---
```{r}
knitr::opts_chunk$set(echo = T, message = FALSE, warning = FALSE)
stdt<-date()
```

# Navigate to data directory
```{bash}
DATADIR=/hpc/group/chsi-mic-2022/data
cd ${DATADIR}
pwd
ls -l
```

# BCL files

Get tar of bcl files and unzip them
```{bash}
#wget -P /hpc/group/chsi-mic-2022/data https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-1.2.0.tar.gz\
#tar -zxvf /hpc/group/chsi-mic-2022/data/cellranger-tiny-bcl-1.2.0.tar.gz
```

# Get sample sheet
```{bash}
#mkdir cellranger-tiny-csv-1.2.0/
#wget --directory-prefix=cellranger-tiny-csv-1.2.0/ https://cf.10xgenomics.com/supp/cell-exp/cellranger-tiny-bcl-simple-1.2.0.csv
```

# Get fastqs for the pbmc8k dataset
```{bash}
#mkdir pbmc8k
#wget --directory-prefix=pbmc8k/ http://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/2.1.0/pbmc8k/pbmc8k_fastqs.tar
```

# Run Cellranger

Make sure that the cellranger works from within the course singularity container
```{bash}
cellranger --version
```

**NOTE!**

- I haven't actually run the cellranger commands in the course workspace, I just copied /outs from plp server and updated the paths above
- We would need to a copy #ANNO=/mnt/data2/Annotation/CellRanger/refdata-gex-GRCh38-2020-A/ to the workspace


Run cellranger's mkfastq command to reformat BCL files into fastqs
```{bash eval = FALSE}
DATADIR=/hpc/group/chsi-mic-2022/data
BCLS=${DATADIR}/cellranger-tiny-bcl-1.2.0
SAMP=${BCLS}/cellranger-tiny-bcl-simple-1.2.0.csv
FASTQ=${DATADIR}/pbmc8k/fastq
cellranger mkfastq \
    --run=${BCLS} \
    --csv=${SAMP} \
    --output-dir=${FASTQ}
```

Run cellranger's count command to map reads to the reference transcriptome and count the reads per feature
```{bash eval=FALSE}
#ANNO=/mnt/data2/Annotation/CellRanger/refdata-gex-GRCh38-2020-A/
cellranger count \
    --id=pbmc8k \
    --sample=pbmc8k \
    --fastqs ${FASTQ} \
    --transcriptome=${ANNO} \
    --localcores 20
```
